Add ability hash all assets output by webpack

ostowe opened the pull request:
- Switch from using filename and chunkFilename templates for hash replacement to leveraging compiler events to access the asset files directly.
- Alter surrounding code to support the updated hash replacement logic including updated regex builder, tests

ostowe committed:
Add emit handler instead of explicit filename and chunkFilename loops

ostowe committed:
Fix regex logic to work with new method of replacing git hash

ostowe committed:
Alter readme for new regexes, update tests

ostowe committed:
Bump version

ostowe commented on index.js:
```diff
@@ -60,25 +62,56 @@ function WebpackGitHash(opts) {
-  if (err) {
-    throw err;
+WebpackGitHash.prototype.populateRegex = function(assetName) {
+  if (!this.regex.hasOwnProperty(assetName)) {
```
hmm woops, forgot to change this, I might switch back to using an object for `this.regex` so we can cache the regex builds.

ostowe committed:
Switch back to using an object for this.regex

- Fix a bug with assetNames

ostowe commented:
I'm actually not sure this will work at all. It appears to confuse the stats object webpack produces after it's finished. Taking a look to see if I can figure it out.

ostowe committed:
Add logic to replace chunk files with updated asset names

- Update test correspondingly

ostowe commented:
@joshkadis any comments here? have a chance to take a look?

Commit:
update README

Commit:
don't need this line

ostowe committed:
Fix file deletion and callback functionality

- Add .map to all regexes to cut down on # of regexes produced
- Only test files that actually have [githash] placeholder in them
  (instead of everyithing!)
- Switch to using full filepath for regex matching instead of trying to replace output path

ostowe committed:
Merge branch 'feature/hash-everything' of github.com:ostowe/webpack-git-hash into feature/hash-everything

joshkadis commented on index.js:
```diff
@@ -85,6 +81,10 @@ WebpackGitHash.prototype.deleteObsoleteFile = function(file) {
  */
 WebpackGitHash.prototype.populateRegex = function(assetName) {
+  // Sourcemap extension is included in regex
+  assetName = assetName.replace('.map', '');
```
Use a regex here to make sure we only replace `.map` at the end of the string

```
assetName.replace(/\.map$/, '');
```

joshkadis commented on index.js:
```diff
@@ -122,10 +122,11 @@ WebpackGitHash.prototype.replaceAsset = function(compilation, assetName) {
+    console.log('WebpackGitHash: hash added to ' + assetName);
+
+    this.populateRegex(assetName);
+  }
```
probably worth logging for skipped files too, 

```
  } else {
    console.log('WebpackGitHash: skipped ' + assetName);
  }
```

joshkadis commented on index.js:
```diff
@@ -161,9 +162,12 @@ WebpackGitHash.prototype.buildRegex = function(template, hash) {
 
-  // Filename must come at end of string to avoid `filename.css` matching `filename.css.map`
+  // remove sourcemap extension, use in regex instead
+  regex = regex.replace('.map', '');
```
same as above re: regex to make sure we don't replace map in the middle of a file name

ostowe committed:
Ensure .map is only being replaced at end of filename

- Log when hash is not added to a filename (skipped)

ostowe committed:
Switch back to swapping placeholder in filename and chunkFilename

- Compiled files still contained references to placeholder, needd to
  replace earlier
- Maintain asset and chunk asset replacement for extract text plugin
  assets

deiwin commented:
What's the status of this PR? Anything stopping this from being merged?

deiwin commented:
I had to explicitly install the `fs-extra` package to get the code from this PR to work.
